networkx/algorithms/community/demon.py

```diff
@@ -0,0 +1,296 @@
+                        if weight:
+                            label_freq[nn_c] = label_freq[nn_c] + ego_minus_ego.edge[nn][n][weight]
+                        else:
+                            label_freq[nn_c] += 1
+                    else:
+                        # case of weighted graph
+                        if weight:
+                            label_freq[nn_c] = ego_minus_ego.edge[nn][n][weight]
+                        else:
+                            label_freq[nn_c] = 1
+
+            # first run, random choosing of the communities among the neighbors labels
+            if t == 0:
+                if not len(n_neighbors) == 0:
+                    r_label = random.sample(label_freq.keys(), 1)
+                    ego_minus_ego.node[n]['communities'] = r_label
+                    old_node_to_coms[n] = r_label
+
+            # choosing the majority
+            else:
+                max_freq = max(label_freq.values())
+                labels = [l for l in label_freq if label_freq[l] == max_freq]
+
+                node_to_coms[n] = labels
+
+                if n not in old_node_to_coms or not set(node_to_coms[n]) == set(old_node_to_coms[n]):
+                    old_node_to_coms[n] = node_to_coms[n]
+                    ego_minus_ego.node[n]['communities'] = labels
+
+    # build the communities reintroducing the ego
+    community_to_nodes = {}
+    for n in nx.nodes(ego_minus_ego):
+        if len(nx.neighbors(ego_minus_ego, n)) == 0:
+            ego_minus_ego.node[n]['communities'] = [n]
+
+        c_n = ego_minus_ego.node[n]['communities']
+
+        for c in c_n:
+            if c in community_to_nodes:
+                community_to_nodes[c].append(n)
+            else:
+                community_to_nodes[c] = [n, ego]
+
+    return community_to_nodes
+
+
+def _merge_communities(communities, actual_community, epsilon):
+    """
+    Builds communities using the incomplete communities detected by the label
+    propagation algorithm in the EgoMinusEgo graphs.
+
+    Parameters
+    ----------
+    communities: dict
+        Dictionary of communities
+
+    actual_community: set
+        A community
+
+    epsilon: real
+        The tolerance required in order to merge communities.
+
+    Returns
+    -------
+    communities : list
+        Merged communities
+    """
+
+    # if the community is already present return
+    if actual_community in communities:
+        return communities
+
+    else:
+        # search a community to merge with
+        inserted = False
+
+        for test_community in communities:
+            union = _generalized_inclusion(actual_community, test_community, epsilon)
+            # community to merge with found
+            if union:
+                communities.pop(communities.index(test_community))
+                communities.append(union)
+                inserted = True
+                break
+
+        # not merged: insert the original community
+        if not inserted:
```

Comment:
Believe it or not, this is a perfect opportunity to use Python's for/else construction. Remove the `inserted` variable and just replace this line with `else:`.